npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb

```diff
@@ -10,18 +11,52 @@ def initialize(package_json, lockfiles:)
         @package_json = package_json
         @lockfiles = lockfiles
         @package_manager = package_json.fetch("packageManager", nil)
+        @engines = package_json.fetch("engines", nil)
       end
 
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/PerceivedComplexity
       def setup(name)
-        return unless @package_manager.nil? || @package_manager.start_with?("#{name}@")
+        # we prioritize version mentioned in "packageManager" instead of "engines"
+        # i.e. if { engines : "pnpm" : "6" } and { packageManager: "pnpm@6.0.2" },
+        # we go for the sepcificity mentioned in packageManager (6.0.2)
+        # Dependabot::Experiments.register(:enable_pnpm_yarn_dynamic_engine, true)
+        if Dependabot::Experiments.enabled?("enable_pnpm_yarn_dynamic_engine")
+
+          unless @package_manager.nil? || @package_manager.start_with?("#{name}@") ||
+                 (@package_manager && @package_manager == name.to_s)
+            return
+          end
 
-        version = requested_version(name)
+          if @engines && @package_manager.nil?
+            # if "packageManager" doesn't exists in manifest file,
+            # we check if we can extract "engines" information
+            version = check_engine_version(name)
+
+          elsif @package_manager&.==name.to_s
+            # if "packageManager" is found but no version is specified (i.e. pnpm@1.2.3),
+            # we check if we can get "engines" info to override default version
+            Dependabot.logger.info("Found \"packageManager\" : \"#{@package_manager}\"")
+            version = check_engine_version(name) if @engines
+
+          elsif @package_manager&.start_with?("#{name}@")
+            # if "packageManager" info has version specification i.e. yarn@3.3.1
+            # we go with the version in "packageManager"
+            Dependabot.logger.info("Found \"packageManager\" : \"#{@package_manager}\"")
+          end
+        else
+          return unless @package_manager.nil? || @package_manager.start_with?("#{name}@")
```

Comment:
nit: same as above, use the safe navigation operator

```suggestion
          return unless @package_manager&.start_with?("#{name}@")
```

Reply:
@JamieMagee , i think _nil?_ check needs exclusive condition check. as this condition did already exist (previous dev work), i have not tried to alter it (i tried and it was failing test cases) . I think we can leave it as it is for now.